RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc

```diff
@@ -0,0 +1,628 @@
+   float tmp_x = float(row_offset) + 0.5f;
+   float tmp_y = float(col_offset) + 0.5f;
+
+   
+   //--- Store these offsets (to be added later) in a LocalPoint after tranforming
+   //    them from measurement units (pixel units) to local coordinates (cm)
+   LocalPoint lp;
+   if ( theClusterParam.with_track_angle )
+      //--- Update call with trk angles needed for bow/kink corrections  // Gavril
+      lp = theDetParam.theTopol->localPosition( MeasurementPoint(tmp_x, tmp_y), theClusterParam.loc_trk_pred );
+   else
+   {
+      edm::LogError("PixelCPEClusterRepair")
+      << "@SUB = PixelCPEClusterRepair::localPosition"
+      << "Should never be here. PixelCPEClusterRepair should always be called with track angles. This is a bad error !!! ";
+      lp = theDetParam.theTopol->localPosition( MeasurementPoint(tmp_x, tmp_y) );
+   }
+   
+   //--- Compute the size of the matrix which will be passed to TemplateReco.
+   //    We'll later make  clustMatrix[ mrow ][ mcol ]
+   int mrow=0, mcol=0;
+   for (int i=0 ; i!=theClusterParam.theCluster->size(); ++i )
+   {
+      auto pix = theClusterParam.theCluster->pixel(i);
+      int irow = int(pix.x);
+      int icol = int(pix.y);
+      mrow = std::max(mrow,irow);
+      mcol = std::max(mcol,icol);
+   }
+   mrow -= row_offset; mrow+=1; mrow = std::min(mrow,cluster_matrix_size_x);
+   mcol -= col_offset; mcol+=1; mcol = std::min(mcol,cluster_matrix_size_y);
+   assert(mrow>0); assert(mcol>0);
+
+
+   //--- Make and fill the bool arrays flagging double pixels
+   bool xdouble[mrow], ydouble[mcol];
+   // x directions (shorter), rows
+   for (int irow = 0; irow < mrow; ++irow)
+      xdouble[irow] = theDetParam.theRecTopol->isItBigPixelInX( irow+row_offset );
+   //
+   // y directions (longer), columns
+   for (int icol = 0; icol < mcol; ++icol)
+      ydouble[icol] = theDetParam.theRecTopol->isItBigPixelInY( icol+col_offset );
+
+   //--- C-style matrix.  We'll need it in either case.
+   float clustMatrix[mrow][mcol];
+   float clustMatrix2[mrow][mcol];
+
+   //--- Prepare struct that passes pointers to TemplateReco.  It doesn't own anything.
+   SiPixelTemplateReco::ClusMatrix   clusterPayload  { &clustMatrix[0][0], xdouble, ydouble, mrow,mcol};
+   SiPixelTemplateReco2D::ClusMatrix clusterPayload2d{ &clustMatrix2[0][0], xdouble, ydouble, mrow,mcol};
+
+
+   //--- Copy clust's pixels (calibrated in electrons) into clustMatrix;
+   memset( clustMatrix, 0, sizeof(float)*mrow*mcol );   // Wipe it clean.
+   for (int i=0 ; i!=theClusterParam.theCluster->size(); ++i )
+   {
+      auto pix = theClusterParam.theCluster->pixel(i);
+      int irow = int(pix.x) - row_offset;
+      int icol = int(pix.y) - col_offset;
+      // &&& Do we ever get pixels that are out of bounds ???  Need to check.
+      if ( (irow<mrow) & (icol<mcol) ) clustMatrix[irow][icol] =  float(pix.adc);
+   }
+   // fillClustMatrix( float * clustMatrix );
```

Comment:
Is this commented out lines needed? If so, please add another line of comment which explain why; otherwise just remove it

Reply:
This code will change.  I'd like to keep the commented line for now, since I may uncomment it in the near future.